lib/jormungandr/src/Cardano/Wallet/Jormungandr/Network.hs

```diff
@@ -0,0 +1,147 @@
+-- mkNetworkLayer _httpBridge = NetworkLayer
+--     { nextBlocks = undefined
+--     , networkTip = undefined
+--     , postTx = undefined
+--     }
+--
+-- -- | Creates a jormungandr 'NetworkLayer' using the given connection
+-- -- settings.
+-- newNetworkLayer
+--     :: Int -> IO (NetworkLayer IO)
+-- newNetworkLayer port = mkNetworkLayer <$> newHttpBridge port
+--   where
+--     newHttpBridge = undefined
+
+{-------------------------------------------------------------------------------
+                            Jormungandr Client
+-------------------------------------------------------------------------------}
+
+-- | Endpoints of the jormungandr REST API.
+newtype Jormungandr m = Jormungandr
+    { getTipId
+        :: ExceptT ErrNetworkUnreachable m BlockId
+    }
+
+-- | Construct a 'Jormungandr'-client
+--
+-- >>> mgr <- newManager defaultManagerSettings
+-- >>> j = mkJormungandr mgr (BaseUrl Http "localhost" 8080 "")
+-- >>> runExceptT $ getTipId j
+-- Right (BlockId (Hash {getHash = "26c640a3de09b74398c14ca0a137ec78"}))
+mkJormungandr
+    :: Manager -> BaseUrl -> Jormungandr IO
+mkJormungandr mgr baseUrl = Jormungandr
+    { getTipId = ExceptT $ do
+        let ctx = safeLink api (Proxy @GetTipId)
+        run cGetTipId >>= defaultHandler ctx
+    }
+  where
+    run :: ClientM a -> IO (Either ServantError a)
+    run query = runClientM query (mkClientEnv mgr baseUrl)
+
+    defaultHandler
+        :: Link
+        -> Either ServantError a
+        -> IO (Either ErrNetworkUnreachable a)
+    defaultHandler ctx = \case
+        Right c -> return $ Right c
+
+        -- The node has not started yet or has exited.
+        -- This could be recovered from by either waiting for the node
+        -- initialise, or restarting the node.
+        Left (ConnectionError e) ->
+            return $ Left $ ErrNetworkUnreachable e
+
+        -- Other errors (status code, decode failure, invalid content type
+        -- headers). These are considered to be programming errors, so crash.
+        Left e -> do
+            throwM (ErrUnexpectedNetworkFailure ctx e)
+
+    cGetTipId = client api
+
+data ErrUnexpectedNetworkFailure
+    = ErrUnexpectedNetworkFailure Link ServantError
+    deriving (Show)
+
+instance Exception ErrUnexpectedNetworkFailure
+
+-- | Tries to waits 20 s, until 'getTipId jormungandr' succeeds.
+--
+-- NOTE: There is a similar helper for 'NetworkLayer' in
+-- "Cardano.Wallet.Network".
+waitForConnection
+    :: Jormungandr IO
+    -> Int -- ^ Number of retries / timeout in seconds
+    -> IO ()
+waitForConnection j = loop
+  where
+    loop :: Int -> IO ()
+    loop retries = runExceptT (getTipId j) >>= \case
+        Right _ -> do
+            return ()
+        Left e@(ErrNetworkUnreachable _) | retries > 0 -> do
+                     TIO.putStrLn "[INFO] waiting for connection to the node..."
+                     threadDelay 1000000
+                     loop (retries - 1)
+               | otherwise -> throwIO e
```

Comment:
This function already exists in `Cardnao.Wallet.Network`. What's the rationale for having both?

Reply:
To have it for both `Jormungandr`, and `NetworkLayer`. (Ignoring HttpBridge as I haven't seen a need for it)

https://github.com/input-output-hk/cardano-wallet/pull/321#discussion_r290200068

Reply:
But what's the need for the one in Jörmungandr? Why isn't it possible to get the same "feature" using only the `NetworkLayer` ? What does it bring us to specialize this to `Jormungandr`?

Reply:
We _could_ decide to _not_ have integration tests for `Jormungandr`. Then we wouldn't need a `waitForConnection`. I don't see why we shouldn't have though.

If we instead were to here implement and test `networkTip networkLayer`, that would rely on both `getTipId` (implemented) and `getBlock` (not implemented), as `getTipId` doesn't return a `BlockHeader`.